File: functions/src/models/database/section_data.ts
export class SectionData implements DatabaseEntry {
  readonly path: string;
  readonly data: Record<string, unknown>;

  constructor(data?: {name: string, folderId: string, useCasesDocId: string}) {
    this.path = 'sections';
    this.data = data ?? {};
  }

  set name(name: string) {
    this.data['name'] = name;
  }

  set folderId(folderId: string) {
    this.data['folderId'] = folderId;
  }

  set useCasesDocId(useCasesDocId: string) {
    this.data['useCasesDocId'] = useCasesDocId;
  }
}

File: shared_models/lib/src/database_entries/section_data.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:googleapis/firestore/v1.dart';
import 'package:shared_models/src/database_entries/database_entry_mixin.dart';

part 'section_data.freezed.dart';
part 'section_data.g.dart';

@freezed
abstract class SectionData with _$SectionData, DatabaseEntry {
  factory SectionData(
      {required String name,
      String? folderId,
      String? useCasesDocId}) = _SectionData;

  factory SectionData.fromJson(Map<String, dynamic> json) =>
      _$SectionDataFromJson(json);

  @override
  Document toDocument() {
    return Document()
      ..fields = {
        'name': Value()..stringValue = name,
        'folderId': Value()..stringValue = folderId,
        'useCasesDocId': Value()..stringValue = useCasesDocId
      };
  }
}
